### PR TITLE
Report VM metrics using libvirt APIs

### PR DESCRIPTION
**Pull request by fromanirh**
**What this PR does / why we need it**:
This PR introduces the VM metrics reporting from virt-handler, leveraging the existing prometheus endpoint added circa version 0.10.0.

Three patches are presented here:
1. the first patch introduces support for the GetDomainStats() virtwrap API. We add new datatypes which closely mimic the libvirt bulk stats (https://godoc.org/github.com/libvirt/libvirt-go#DomainStats) and its sub-types.
   We decide to replicate the libvirt types because the DomainStats type include a pointer to a libvirt.Domain, something we don't want to have as part of the API. Considering the surrounding code, it seems preferible not to expose native libvirt types at the API level of the virtwrap module, and this provide the second part of the rationale to replicate them.
  Internally, the implementation uses the GetAllDomainStats API, which is the modern recommended way to grab the libvirt metrics.
   Finally, we also report memory metrics using the MemoryStats API. A convenient subtype which mimic the bulk stats API types is also added.
2. the second patch exposes the GetDomainStats through the cmd-client socket protocol. With this patch, clients like virt-handler can consume the stats retrieved by virt-launcher.
3. the third patch implement very basic logic in virt-handler to collect the metrics using the cmd-client socket and expose them through the existing prometheus endpoint.
The current approach is a poll loop, implemented for the sake of simplicity.

**Fixes**:
N/A 

**Special notes for your reviewer**:

There are some design items which should be discussed here:
1. should the libvirt types be replicated or can we exposes them directly as part of the API?
2. is virt-handler the right place to expose the metrics to prometheus?
3. is an internal polling acceptable or should we use some smarter approach?
4. we know from the experience from oVirt that libvirt bulk stats call may block, and we can't do anything about that. How should that be handled in kubevirt?

```release-note
Exposes VM metrics to prometheus
```

### CODE REVIEW

**Comment by fabiand** [comment]
ci test please

File: cmd/virt-handler/virt-handler.go [hunk 1]
```diff
@@ -182,6 +183,10 @@ func (app *virtHandlerApp) Run() {
 	if err != nil {
 		glog.Fatalf("unable to generate certificates: %v", err)
 	}
+
+	updater := metrics.NewUpdater(app.VirtShareDir, 10*time.Second)
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply cmd/virt-handler/virt-handler.go]: Thanks @rmohr for the hint, this is indeed the hint I was looking for. I'll explore this direction and update the PR accordingly.

**Comment by fromanirh** [comment]
@rmohr I replaced the polling loop. I think makes little sense to update this PR adding the polling and replacing one patch later -we don't want the polling at all, don't we?

Thus I sent https://github.com/kubevirt/kubevirt/pull/1840. Please note that most of the content is identical between the two PRs. Only one commit (https://github.com/kubevirt/kubevirt/pull/1813/commits/3c6ff2843b46bc0c1aa9bfcb5b281d3ee30bdd08) is being replaced.

Let me know whic PR you want to keep, I'll close the other one.

**Comment by fromanirh** [comment]
> @rmohr I replaced the polling loop. I think makes little sense to update this PR adding the polling and replacing one patch later -we don't want the polling at all, don't we?
> 
> Thus I sent #1840. Please note that most of the content is identical between the two PRs.
> 
> The only difference is that the [polling loop](https://github.com/kubevirt/kubevirt/pull/1813/commits/3c6ff2843b46bc0c1aa9bfcb5b281d3ee30bdd08) is replaced by a [collector](https://github.com/kubevirt/kubevirt/pull/1840/commits/afb5109d1c7262986f4a929f55fd2244dd74885c).
> 
> Let me know whic PR you want to keep, I'll close the other one.

/cc @vladikr @davidvossel @fabiand @slintes

**Comment by fromanirh** [comment]
let's continue on https://github.com/kubevirt/kubevirt/pull/1840

**Pull request closed without merging**